### PR TITLE
Disability increase view field

### PR DESCRIPTION
**Pull request by cehsu**
These changes set up a `VerifiedReviewContainer` and `VerifiedReviewPage` components (based largely on the `ReviewPage` and `ReviewCollapsibleChapter` components, respectively) for the 526 disability increase application.

[Prototype link](https://adhoc.invisionapp.com/share/9MFXXQTUGDK#/screens/282992016).

Review
localhost_3001_disability-benefits_526_apply-for-increase_

Update
localhost_3001_disability-benefits_526_apply-for-increase_ 1

View (chapter/page not included in form yet)
localhost_3001_disability-benefits_526_apply-for-increase_introduction 6

### CODE REVIEW

File: src/js/common/schemaform/save-in-progress/reducers.js [hunk 1]
```diff
@@ -68,7 +68,10 @@ export const saveInProgressReducers = {
 
     // if we’re prefilling, we want to use whatever initial data the form has
     if (state.prefillStatus === PREFILL_STATUSES.pending) {
-      const formData = _.merge(state.data, action.data.formData);
+      let formData = _.merge(state.data, action.data.formData);
+      if (state.storePrefillStatus) {
+        formData = _.merge(state.data, action.data.formData, { prefilled: true });
+      }
```
Comment (cehsu) [inline src/js/common/schemaform/save-in-progress/reducers.js]: `prefillStatus` will be stored to ensure that the applicant is presented with a consistent experience if they save their form and log in again later to complete it.

File: src/js/disability-benefits/526EZ/components/ReviewCollapsiblePage.jsx [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+            onSubmit={({ formData }) => this.handleSubmit(formData, page.pageKey, page.arrayPath, page.index)}
+            onChange={(formData) => this.onChange(formData, page.arrayPath, page.index)}
+            uploadFile={this.props.uploadFile}
+            reviewMode={!editing}
+            formContext={formContext}
+            editModeOnReviewPage={page.editModeOnReviewPage}>
+            {!editing ? <div/> : <ProgressButton
+              submitButton
+              buttonText="Update Page"
+              buttonClass="usa-button-primary"/>}
```
Comment (cehsu) [inline src/js/disability-benefits/526EZ/components/ReviewCollapsiblePage.jsx]: The prototype depicts a "Cancel" button, but that seems like it would involve disabling autosave, so it was not provided here.

File: src/js/disability-benefits/526EZ/pages/veteranInfo.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        type: 'string',
+        pattern: '^[cC]{0,1}\\d{7,9}$'
+      },
+      gender: {
+        type: 'string',
+        'enum': ['F', 'M']
+      },
+      dateOfBirth: date
+    }
+  };
```
Comment (cehsu) [inline src/js/disability-benefits/526EZ/pages/veteranInfo.js]: Most of these schema definitions will be replaced with property references once the backend schema is updated.

**Comment by jbalboni** [comment]
@cehsu I definitely want to find some time to talk about this tomorrow or Friday. There are a lot of changes to the shared form code in here, which I'd like to avoid if it all possible. I want to better understand the requirements and get a good handle on the reasoning for everything in here.

File: src/js/disability-benefits/526EZ/helpers.jsx [hunk 4]
```diff
@@ -30,13 +37,27 @@ export function flatten(data) {
...
+  if (state.prefilStatus === PREFILL_STATUSES.success) {
+    newData.prefilled = true;
+  }
+
+  return {
+    metadata,
+    formData: newData,
+    pages
+  };
+}
```
Comment (cehsu) [inline src/js/disability-benefits/526EZ/helpers.jsx]: It might make sense to eventually include this in the metadata.

File: src/js/disability-benefits/526EZ/helpers.jsx [hunk 5]
```diff
@@ -250,7 +271,7 @@ const listifyCenters = (center, idx, list) => {
   const atLeastThree = list.length > 2;
   return (
     <span key={idx}>
-      {!notLast && !justOne && <span className="repose"> and </span>}
+      {!notLast && !justOne && <span className="unstyled-word"> and </span>}
```
Comment (cehsu) [inline src/js/disability-benefits/526EZ/helpers.jsx]: Fixed a vague classname here.

File: src/js/disability-benefits/526EZ/pages/veteranInfo.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import currentOrPastDateUI from '../../../common/schemaform/definitions/currentOrPastDate';
+import fullNameUI from '../../../common/schemaform/definitions/fullName';
+import ssnUI from '../../../common/schemaform/definitions/ssn';
+import { genderLabels } from '../../../common/utils/labels';
+
+import { veteranInformationViewField, getVerifiedChapterPair } from '../helpers';
+
+export default function createVeteranInfoChapter(formSchema, isReview) {
+
+  const { fullName, date } = formSchema.definitions;
```
Comment (cehsu) [inline src/js/disability-benefits/526EZ/pages/veteranInfo.js]: Should ssn, gender, vaFileNumber be added to the backend schema?

File: src/js/common/schemaform/save-in-progress/SaveInProgressIntro.jsx [hunk 7]
```diff
@@ -93,8 +94,16 @@ export default class SaveInProgressIntro extends React.Component {
     return alert;
   }
 
+  getStartPage = () => {
+    const { form, route, location } = this.props;
+    const formData = form && form.data || {};
+    if (route && location) return getNextPagePath(route.pageList, formData, location.pathname);
+    return this.props.pageList[1].path;
+  }
```
Comment (cehsu) [inline src/js/common/schemaform/save-in-progress/SaveInProgressIntro.jsx]: The idea here is to only calculate the first page if the first page is conditional.

File: src/js/disability-benefits/526EZ/pages/veteranInfo.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+      'ui:title': 'Gender',
+      'ui:options': {
+        labels: genderLabels
+      },
+    }
+  };
+
+  const schema = {
+    type: 'object',
+    required: ['fullName', 'socialSecurityNumber', 'vaFileNumber', 'gender', 'dateOfBirth'],
```
Comment (cehsu) [inline src/js/disability-benefits/526EZ/pages/veteranInfo.js]: @goldenmeanie @cvalarida @markgreenburg can you provide guidance on what is required here? The prototype indicates that nothing is required (perhaps due to the prefill), the request model and backend schema don't seem to mention these.

File: src/js/disability-benefits/526EZ/pages/veteranInfo.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+  const uiSchema = {
+    fullName: fullNameUI,
+    socialSecurityNumber: _.assign(ssnUI, {
+      'ui:title': 'Social Security number',
+      'ui:required': form => !form.vaFileNumber,
+    }),
+    vaFileNumber: {
+      'ui:title': 'VA file number',
+      'ui:required': form => !form.socialSecurityNumber,
```
Comment (cehsu) [inline src/js/disability-benefits/526EZ/pages/veteranInfo.js]: This is how we conditionally require ssn/vaFileNumber on other forms, depending on what is required and how, this may need to be updated.

Comment (goldenmeanie) [reply src/js/disability-benefits/526EZ/pages/veteranInfo.js]: Should match that conditional behavior from other forms

File: src/js/disability-benefits/526EZ/components/ReviewCollapsiblePage.jsx [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+import _ from 'lodash/fp';
+import classNames from 'classnames';
+
+import { focusElement } from '../../../common/utils/helpers';
+import SchemaForm from '../../../common/schemaform/components/SchemaForm';
+import ProgressButton from '../../../common/components/form-elements/ProgressButton';
+
+const Element = Scroll.Element;
+const scroller = Scroll.scroller;
+export default class ReviewCollapsiblePage extends React.Component {
```
Comment (jbalboni) [inline src/js/disability-benefits/526EZ/components/ReviewCollapsiblePage.jsx]: I think we should name this something else now. It's not collapsible

File: src/js/disability-benefits/526EZ/components/VerifiedReviewPage.jsx [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+  render() {
+    const { form, user, formContext, route } = this.props;
+    const { chapterKey, pageKey } = route.pageConfig;
+    const page = formConfig.chapters[chapterKey].pages[pageKey];
+    const isNotPrefilled = !form.data.prefilled;
+
+    if (isNotPrefilled) return <RoutedSavablePage {...this.props}/>;
+    return (
+      <div>
+        <p>Please review the information we have on file for you. If something doesn’t look right, you can fix it by clicking the Edit button.</p>
```
Comment (jbalboni) [inline src/js/disability-benefits/526EZ/components/VerifiedReviewPage.jsx]: Could this text go in ReviewCollapsiblePage? I'm thinking that eventually we want to get to where we could wrap your custom ReivewCollapsiblePage component in RoutedSavablePage and reduce some of the duplication of SiP code. I don't think that's necessary right now, but I'd want to do that before making this a recommended pattern for forms.

File: src/js/disability-benefits/526EZ/config/form.js [hunk 12]
```diff
@@ -76,49 +79,26 @@ const formConfig = {
   formId: '21-526EZ',
   version: 1,
   migrations: [],
+  prefillTransformer,
   savedFormMessages: {
     notFound: 'Please start over to apply for disability claims increase.',
     noAuth: 'Please sign in again to resume your application for disability claims increase.'
   },
+  storePrefillStatus: true,
```
Comment (jbalboni) [inline src/js/disability-benefits/526EZ/config/form.js]: Is this necessary now?

File: src/js/disability-benefits/526EZ/helpers.jsx [hunk 13]
```diff
@@ -30,13 +37,27 @@ export function flatten(data) {
 
 export function transform(formConfig, form) {
   const formData = flatten(transformForSubmit(formConfig, form));
+  delete formData.storePrefillStatus;
```
Comment (jbalboni) [inline src/js/disability-benefits/526EZ/helpers.jsx]: I think you'd want to delete formData.prefilled now?

File: src/js/disability-benefits/526EZ/helpers.jsx [hunk 14]
```diff
@@ -30,13 +37,27 @@ export function flatten(data) {
...
   return JSON.stringify({
     educationBenefitsClaim: {
       form: formData
     }
   });
 }
 
+export function prefillTransformer(pages, formData, metadata, state) {
+  const newData = cloneDeep(formData);
```
Comment (jbalboni) [inline src/js/disability-benefits/526EZ/helpers.jsx]: You shouldn't need to deep clone this. I think you can use `set` from data-utils on line 52 and set newData to the result. It'll do a shallow clone for you.

File: src/js/disability-benefits/526EZ/helpers.jsx [hunk 15]
```diff
@@ -283,3 +304,105 @@ export const evidenceSummaryView = ({ formData }) => {
...
+};
+
+const getVerifiedPairPagePaths = (chapterTitleWords) => {
+  const unverifiedPagePath = kebabize(chapterTitleWords);
+  chapterTitleWords.unshift('review');
+  const verifiedPagePath = kebabize(chapterTitleWords);
+  return { unverifiedPagePath, verifiedPagePath };
+};
+
+export const getVerifiedChapterPair = (chapterConfig) => {
```
Comment (jbalboni) [inline src/js/disability-benefits/526EZ/helpers.jsx]: I think I understand what this function is doing, but the word "pair" makes me think I'm missing something. You might consider adding some comments around that.

File: src/js/common/schemaform/review/ObjectField.jsx [hunk 16]
```diff
@@ -1,6 +1,7 @@
 import PropTypes from 'prop-types';
 import React from 'react';
 import _ from 'lodash/fp';
+import classNames from 'classnames';
```
Comment (jbalboni) [inline src/js/common/schemaform/review/ObjectField.jsx]: Are the changes in here still necessary?

File: src/js/disability-benefits/526EZ/components/IntroductionPage.jsx [hunk 17]
```diff
@@ -17,6 +17,7 @@ class IntroductionPage extends React.Component {
         <FormTitle title="This page needs attention"/>
         <p>Equal to VA Form 21-526EZ.</p>
         <SaveInProgressIntro
+          {...this.props}
```
Comment (jbalboni) [inline src/js/disability-benefits/526EZ/components/IntroductionPage.jsx]: Is this change still necessary?

**Review by jbalboni — COMMENTED** [review]
Looks good so far! I had some questions/comments.

File: src/js/disability-benefits/526EZ/helpers.jsx (hunk 15 above)
Comment (cehsu) [reply src/js/disability-benefits/526EZ/helpers.jsx]: Good point, it should be updated to getChapter (pair will be removed throughout). `createVeteranInfoChapter`should be curried so that `isReview` doesn't need to be supplied. Will update.

File: src/js/disability-benefits/526EZ/pages/veteranInfo.js (hunk 8 above)
Comment (cehsu) [reply src/js/disability-benefits/526EZ/pages/veteranInfo.js]: Thanks @goldenmeanie , sounds good. We'll make ssn conditionally required if VAFileNumber is not present (and vice versa), according to your comment below.

File: src/js/disability-benefits/526EZ/pages/veteranInfo.js (hunk 9 above)
Comment (cehsu) [reply src/js/disability-benefits/526EZ/pages/veteranInfo.js]: Thanks @goldenmeanie , in that case, shall we update the title to read:
ssn:
'Social Security number (must have this or a VA file number)'
vaFileNumber:
'VA file number (must have this or a Social Security number)'

File: src/js/disability-benefits/526EZ/components/ReviewCollapsiblePage.jsx (hunk 10 above)
Comment (cehsu) [reply src/js/disability-benefits/526EZ/components/ReviewCollapsiblePage.jsx]: Great point, will rename this `verifiedReviewPage`, and rename `verifiedReviewPage`, `verifiedReviewContainer`.

File: src/js/disability-benefits/526EZ/components/VerifiedReviewPage.jsx (hunk 11 above)
Comment (cehsu) [reply src/js/disability-benefits/526EZ/components/VerifiedReviewPage.jsx]: Sounds good. Updated.

File: src/js/disability-benefits/526EZ/config/form.js (hunk 12 above)
Comment (cehsu) [reply src/js/disability-benefits/526EZ/config/form.js]: That is zombie code now, thanks for catching it. Removed.

File: src/js/disability-benefits/526EZ/helpers.jsx (hunk 13 above)
Comment (cehsu) [reply src/js/disability-benefits/526EZ/helpers.jsx]: Correct, fixed.

File: src/js/disability-benefits/526EZ/pages/veteranInfo.js (hunk 9 above)
Comment (goldenmeanie) [reply src/js/disability-benefits/526EZ/pages/veteranInfo.js]: What do you think about adapting this pattern https://staging.vets.gov/education/apply-for-education-benefits/application/1990E/applicant/information?

**State 1**:
Social Security number {SSN field, **required**}
[ ] I don’t have a Social Security number

VA file number {text field}

**State 2**:
Social Security number {SSN field}
[**×**] I don’t have a Social Security number

VA file number {text field, **required**}

File: src/js/disability-benefits/526EZ/helpers.jsx (hunk 14 above)
Comment (cehsu) [reply src/js/disability-benefits/526EZ/helpers.jsx]: Great idea, thanks @jbalboni, changed.

File: src/js/common/schemaform/review/ObjectField.jsx (hunk 16 above)
Comment (cehsu) [reply src/js/common/schemaform/review/ObjectField.jsx]: Nice catch, removed.

File: src/js/disability-benefits/526EZ/components/IntroductionPage.jsx (hunk 17 above)
Comment (cehsu) [reply src/js/disability-benefits/526EZ/components/IntroductionPage.jsx]: `<SaveInProgressIntro/>` now needs `form`, `route`, `location`, and `pageList `from `this.props` to calculate the first page, should we enumerate them instead

File: src/js/disability-benefits/526EZ/pages/veteranInfo.js (hunk 3 above)
Comment (cehsu) [reply src/js/disability-benefits/526EZ/pages/veteranInfo.js]: `fullName` and `date` are now used, should we add any of the others to the backend schema?

File: src/js/disability-benefits/526EZ/helpers.jsx (hunk 4 above)
Comment (cehsu) [reply src/js/disability-benefits/526EZ/helpers.jsx]: Looping in @egarreau for additional thoughts on this.

File: src/js/disability-benefits/526EZ/chapters/veteranInfo.js [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+      },
+      dateOfBirth: date
+    }
+  };
+
+  const chapterConfig = {
+    chapterTitle: 'Veteran Information',
+    isReview,
+    component: VerifiedReviewContainer,
+    verifiedReviewComponent: veteranInformationViewField,
```
Comment (cehsu) [inline src/js/disability-benefits/526EZ/chapters/veteranInfo.js]: This is now specified here as doing so in helpers.js resulted in the following error: `TypeError: Cannot read property 'default' of undefined`.

File: src/js/disability-benefits/526EZ/components/IntroductionPage.jsx (hunk 17 above)
Comment (cehsu) [reply src/js/disability-benefits/526EZ/components/IntroductionPage.jsx]: Thanks @jbalboni, great idea, updated.

**Review by cvalarida — COMMENTED** [review]
I'm not sure that creating another page component is the right long-term solution to this problem, but we can re-think this when we have more time.

To be honest, I didn't go through this very thoroughly because it's so long, but it looks okay where it's at now. I'm just not sure about how we determine the pre-fill status. Also, I need to fire it up and run through it before I can give it the :+1: ; I'll do that either later today or tomorrow.

**Review by peggygannon — COMMENTED** [review]
@cehsu , please make the following edit to this screen: 
for the first line, please tweak to: 
Please review the information we have on file for you. If something doesn't look right, you can click the Edit button to fix it.

**Comment by cehsu** [comment]
Thanks @peggygannon , updated:
screen shot 2018-04-13 at 4 56 59 pm

**Comment by peggygannon** [comment]
@cehsu Ok, looks good.

File: test/disability-benefits/526EZ/helpers.unit.spec.jsx [hunk 19]
```diff
@@ -1,8 +1,12 @@
...
-import { flatten } from '../../../src/js/disability-benefits/526EZ/helpers.jsx';
+import {
+  flatten,
+  isPrefillDataComplete,
+  prefillTransformer
+} from '../../../src/js/disability-benefits/526EZ/helpers.jsx';
 import initialData from './schema/initialData.js';
 
-const formData = initialData;
+delete initialData.prefilled;
```
Comment (cvalarida) [inline test/disability-benefits/526EZ/helpers.unit.spec.jsx]: This may cause problems for some other tests. May not need to worry about it for now, but if we get some weird failures later on down the road, this might be why.

JS is weird. :man_shrugging:

**Review by cvalarida — APPROVED** [review]
I think we can solve this in a cleaner way by using a custom field, but we can save that for another day since we need this out the door.

I hate being rushed. :disappointed: 

The only major thing I found was the validation: I can go to the review veteran details page, hit edit, remove a required field, and hit continue (without hitting "update page"...which should probably be renamed to "update information" or something).

Other than that, it looks good. I'm approving it because it's not a big deal _yet_, but if you merge without fixing that (for sake of time), please write a ticket to fix that. Bonus points if you write it into the acceptance criteria for the ticket to loop back around with a simpler solution (using a custom field).

File: src/js/disability-benefits/526EZ/helpers.jsx [hunk 20]
```diff
@@ -262,6 +305,100 @@ export const evidenceSummaryView = ({ formData }) => {
...
+
+const SsnViewField = ({ formData }) => {
+  const ssn = formData.slice(5);
+  const mask = <span>•••-••-</span>;
+  return <p>Social Security number: {mask}{ssn}</p>;
+};
+
+const VAFileNumberViewField = ({ formData }) => {
+  const vaFileNumber = formData.slice(5);
+  const mask = <span>•••-••-</span>;
```
[the remark this answers is not in the document]
Comment (cehsu) [reply src/js/disability-benefits/526EZ/helpers.jsx]: Good question, looping in @goldenmeanie , I understand VA file number is typically SSN, but for some older numbers it is different, and we should support both formats.

File: src/js/disability-benefits/526EZ/helpers.jsx [hunk 21]
```diff
@@ -30,13 +37,49 @@ export function flatten(data) {
...
+  return !!(hasVeteranDetails && hasPrimaryAddressInfo && hasPaymentInfo && hasMilitaryHistoryInfo);
+};
+
+export function prefillTransformer(pages, formData, metadata, state) {
+  let newData = formData;
+  const isPrefilled = state.prefilStatus === PREFILL_STATUSES.success;
+  const hasRequiredInformation = isPrefillDataComplete(formData);
+
+  if (isPrefilled && hasRequiredInformation) {
+    newData = set('prefilled', true, newData);
```
[the remark this answers is not in the document]
Comment (cehsu) [reply src/js/disability-benefits/526EZ/helpers.jsx]: Prefill flag marks whether they have info prefilled and also whether they have atleast the required information for each page prefillable, if not, they get rerouted to the unverified view. @goldenmeanie is this reasonable?

File: src/js/disability-benefits/526EZ/config/form.js [hunk 22]
```diff
@@ -83,30 +88,67 @@ const formConfig = {
...
+        }
+        return baseString;
+      },
       pages: {
-        pageOne: {
-          title: 'Page One',
-          path: 'chapter-one/page-one',
-          uiSchema: {},
+        veteranInformation: createVerifiedVeteranInfoPage(fullSchema526EZ),
+        specialCircumstances: { // TODO: create page file and reuse
```
[the remark this answers is not in the document]
Comment (cehsu) [reply src/js/disability-benefits/526EZ/config/form.js]: Special circumstances is used twice, the idea here is to simply DRY this up.

File: src/js/disability-benefits/526EZ/config/form.js [hunk 23]
```diff
@@ -74,6 +78,7 @@ const formConfig = {
   formId: '21-526EZ',
   version: 1,
   migrations: [],
+  prefillTransformer,
```
[the remark this answers is not in the document]
Comment (cehsu) [reply src/js/disability-benefits/526EZ/config/form.js]: Thanks to a reminder courtesy of @jbalboni .

File: test/disability-benefits/526EZ/schema/initialData.js [hunk 24]
```diff
@@ -1,5 +1,34 @@
 export default {
   // For testing purposes only
+  prefilled: true,
```
[the remark this answers is not in the document]
Comment (cehsu) [reply test/disability-benefits/526EZ/schema/initialData.js]: This persists that across multiple sessions, otherwise it's not recorded, and when they return, they would get the unverified view.

File: src/js/disability-benefits/526EZ/components/IntroductionPage.jsx [hunk 25]
```diff
@@ -17,6 +17,7 @@ class IntroductionPage extends React.Component {
         <FormTitle title="This page needs attention"/>
         <p>Equal to VA Form 21-526EZ.</p>
         <SaveInProgressIntro
+          pathname={this.props.location.pathname}
```
[the remark this answers is not in the document]
Comment (cehsu) [reply src/js/disability-benefits/526EZ/components/IntroductionPage.jsx]: This is necessary to determine the conditional first page. since this is the first form that has one.

File: src/js/common/schemaform/save-in-progress/SaveInProgressIntro.jsx [hunk 26]
```diff
@@ -93,8 +95,17 @@ export default class SaveInProgressIntro extends React.Component {
     return alert;
   }
 
+  getStartPage = () => {
```
[the remark this answers is not in the document]
Comment (cehsu) [reply src/js/common/schemaform/save-in-progress/SaveInProgressIntro.jsx]: This determines the first page, which is conditional, for the first time in this form.

File: src/js/common/schemaform/save-in-progress/SaveInProgressIntro.jsx [hunk 27]
```diff
@@ -1,14 +1,16 @@
...
 
+import { getNextPagePath } from '../routing';
 import { toggleLoginModal } from '../../../login/actions';
 import { fetchInProgressForm, removeInProgressForm } from './actions';
 import LoadingIndicator from '@department-of-veterans-affairs/jean-pants/LoadingIndicator';
 import FormStartControls from './FormStartControls';
 import { getIntroState } from './selectors';
 
-export default class SaveInProgressIntro extends React.Component {
+class SaveInProgressIntro extends React.Component {
```
[the remark this answers is not in the document]
Comment (cehsu) [reply src/js/common/schemaform/save-in-progress/SaveInProgressIntro.jsx]: Great point, will update.

File: src/js/disability-benefits/526EZ/helpers.jsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (goldenmeanie) [reply src/js/disability-benefits/526EZ/helpers.jsx]: It can be an SSN or 8 digits with or without a "c" on the end

[the remark this answers is not in the document]
Comment (goldenmeanie) [reply src/js/disability-benefits/526EZ/helpers.jsx]: I think we have a working pattern for this in the EDU forms
